src/lib/models/billing-volume.js

```diff
@@ -172,41 +180,67 @@ class BillingVolume extends Model {
     this._volume = isNull(volume) ? null : parseFloat(volume);
   }
 
+  assertIsNotApproved () {
+    hoek.assert(!this.isApproved, 'Cannot allocate/de-allocate on an approved BillingVolume');
+  }
+
   /**
    * Allocates billing volume
    * @param {Number} ML
    */
   allocate (volume) {
-    validators.assertQuantity(volume);
-    hoek.assert(this.calculatedVolume === this.volume, `Can't allocate ${volume} when volume and calculated volume differ`);
-    this.calculatedVolume = isFinite(this.calculatedVolume) ? this.calculatedVolume + volume : volume;
-    this.volume = this.calculatedVolume;
+    this.assertIsNotApproved();
+    const currentVolume = this.calculatedVolume || new Decimal(0);
+    this.calculatedVolume = currentVolume.add(volume);
+    return this;
   }
 
   /**
    * De-allocate billing volume
    * @param {Number} ML
    */
   deallocate (volume) {
-    hoek.assert(isFinite(this.calculatedVolume), `Can't deallocate ${volume} when calculated volume is not finite`);
-    hoek.assert(this.calculatedVolume === this.volume, `Can't deallocate ${volume} when volume and calculated volume differ`);
-    validators.assertQuantityWithMaximum(volume, this.calculatedVolume);
-    this.calculatedVolume -= volume;
-    this.volume = this.calculatedVolume;
+    this.assertIsNotApproved();
+    const currentVolume = this.calculatedVolume || new Decimal(0);
+    const isVolumeAvailable = currentVolume.greaterThanOrEqualTo(volume);
+    hoek.assert(isVolumeAvailable, `Volume ${volume} cannot be de-allocated, ${currentVolume.toString()} available`);
+    this.calculatedVolume = currentVolume.minus(volume);
+    return this;
   }
 
   /**
-   * Converts volumes to a fixed precision of 3 DP
+   * Sets the volume property from the calculatedVolume decimal
    */
-  toFixed () {
-    if (isFinite(this.volume)) {
-      this.volume = toFixedPrecision(this.volume);
-    }
-    if (isFinite(this.calculatedVolume)) {
-      this.calculatedVolume = toFixedPrecision(this.calculatedVolume);
+  setVolumeFromCalculatedVolume () {
+    const { calculatedVolume } = this;
+
+    this.assertIsNotApproved();
+    if (!assignBillableStatuses.includes(this.twoPartTariffStatus)) {
+      this.volume = is.object(calculatedVolume) && is.directInstanceOf(calculatedVolume, Decimal)
+        ? this.calculatedVolume.toDecimalPlaces(6).toNumber()
+        : this.calculatedVolume;
     }
     return this;
   }
+
+  /**
+   * Gets either the approved volume, or the calculated volume if the billing
+   * volume is not yet approved, as a decimal
+   * @return {Decimal}
+   */
+  get approvedOrCalculatedVolume () {
+    if (this.isApproved) {
+      return isNull(this.volume) ? null : new Decimal(this.volume);
```

Comment:
Wondering if the `null` should be a zero instead? Ignore if not.

Reply:
Null is OK here